drivers: modem: add Telit ME910G1

jeffwelder-ellenbytech opened the pull request:
Added Telit ME910G1 dts and modem chat scripts.

Refactor device macro to be custom for each modem to allow custom configurations on each modem (assisted by @bjarki-trackunit). 
(This was needed by Telit ME910G1 to allow for ON_OFF to be toggled for minimum of 5 seconds as per Telit)

bjarki-trackunit commented:
Looks good :) FYI it will take a while to get merged because of the feature freeze :)

jeffwelder-ellenbytech committed:
drivers: modem: Individualize Modem Device Macros

Refactor device macro to be custom for each modem
to allow custom configurations.

Signed-off-by: Jeff Welder <Jeff.Welder@ellenbytech.com>

jeffwelder-ellenbytech committed:
dts: bindings: Add Telit ME910G1 Modem

Add Telit to vendor-prefixes, and add ME910G1 bindings

Signed-off-by: Jeff Welder <Jeff.Welder@ellenbytech.com>

jeffwelder-ellenbytech committed:
drivers: modem: modem_cellular: add Telit ME910G1

Added chat script and initial config to support Telit ME910G1.
Needed to edit power_pulse pin timing in accordance to datasheet.

Signed-off-by: Jeff Welder <Jeff.Welder@ellenbytech.com>

jeffwelder-ellenbytech committed:
tests: build_all: modem: Add Telit ME910G1 Test Coverage

Added bindings to build_all test for new Telit ME910G1 Modem.

Signed-off-by: Jeff Welder <Jeff.Welder@ellenbytech.com>

bjarki-andreasen reviewed (dismissed):
Nice work :)

bjarki-trackunit commented:
@rerickson1 Could you review this :)

ycsin reviewed (commented):
From the looks of it, it seems that `modem_cellular.c` can grow indefinitely, as the number of supported modem grows?

fabiobaltieri commented on dts/bindings/modem/telit,me910g1.yaml:
```diff
@@ -0,0 +1,23 @@
+# Copyright(c) 2023 Jeff Welder(Ellenby Technologies, Inc.)
+# SPDX - License - Identifier : Apache - 2.0
```
This probably wants to be `SPDX-License-Identifier: Apache-2.0`

rerickson1 replied:
@jeffwelder-ellenbytech please fix this.

bjarki-trackunit commented:
@ycsin The driver is capable of supporing an "infinite" amount of modems, as long as said modems implement the standard AT, CMUX and PPP. However, the driver is an MVP, so it will not get any fancy non generic features, like setting radio access tech, sending sms, internal gnss, etc. so it should not grow in complexity :)

jeffwelder-ellenbytech commented:
Fixed the licensing. 
Working on an improved chat script that works with various eSIMs before merge, will commit again once that is complete.

bjarki-trackunit commented:
> Fixed the licensing. Working on an improved chat script that works with various eSIMs before merge, will commit again once that is complete.

Love it! I have added the DNM (Do Not Merge) label until you merge the fixes, you can remove it and re-request reviews when ready :)

jeffwelder-ellenbytech commented:
I updated the chat scripts to have a more stable experience with the modem. Specifically I was having trouble with Verizon/AT&T eSIMs either giving an error or taking unusually long for setting the APN. 

**Init Script:**
- Added PDP Context Setting (Allows retry on error by restarting init script)

**Dial Script:**
- Remove PDP Context Activation (Note: second parameter not supported by ME910G1 by current firmware)
- Remove PDP Context Setting (Done in init script to allow error retries)
- Added Dummy AT command (Dial script doesn't work if it is only 1 command long)

If the network is too fast to register, the state machine in the dial script will miss the unsolicited registration status, and hang forever. This is fixed in @bjarki-trackunit 's #64659 by polling the registration state which I tested this morning as working. 😄
